
#-##-##-##-##-##-##-##-##-##-##-#

Milestone 3

Nama  : Haafizh Hamda

Batch : FTDS-032-HCK

Program ini dibuat untuk melakukan automatisasi transform dan load data dari PostgreSQL ke ElasticSearch.

#-##-##-##-##-##-##-##-##-##-##-#


In [6]:
# Install the library

# !pip install -q "great-expectations==0.18.19"

In [7]:
# Package
import pandas as pd
from great_expectations.dataset import PandasDataset

In [8]:
# Load
df = pd.read_csv("P2M3_haafizhhamda_data_clean.csv")
gdf = PandasDataset(df)

In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 250 entries, 0 to 249
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   Unnamed: 0         250 non-null    int64 
 1   order_id           250 non-null    object
 2   date               250 non-null    object
 3   product            250 non-null    object
 4   category           250 non-null    object
 5   price              250 non-null    int64 
 6   quantity           250 non-null    int64 
 7   total_sales        250 non-null    int64 
 8   customer_name      250 non-null    object
 9   customer_location  250 non-null    object
 10  payment_method     250 non-null    object
 11  status             250 non-null    object
dtypes: int64(4), object(8)
memory usage: 23.6+ KB


In [10]:
# === REQUIRED EXPECTATIONS ===

# 1 Harus unik
gdf.expect_column_values_to_be_unique("order_id")



{
  "success": true,
  "result": {
    "element_count": 250,
    "missing_count": 0,
    "missing_percent": 0.0,
    "unexpected_count": 0,
    "unexpected_percent": 0.0,
    "unexpected_percent_total": 0.0,
    "unexpected_percent_nonmissing": 0.0,
    "partial_unexpected_list": []
  },
  "meta": {},
  "exception_info": {
    "raised_exception": false,
    "exception_traceback": null,
    "exception_message": null
  }
}

In [11]:
# 2 Nilai di antara rentang wajar
gdf.expect_column_values_to_be_between("quantity", min_value=1, max_value=5)



{
  "success": true,
  "result": {
    "element_count": 250,
    "missing_count": 0,
    "missing_percent": 0.0,
    "unexpected_count": 0,
    "unexpected_percent": 0.0,
    "unexpected_percent_total": 0.0,
    "unexpected_percent_nonmissing": 0.0,
    "partial_unexpected_list": []
  },
  "meta": {},
  "exception_info": {
    "raised_exception": false,
    "exception_traceback": null,
    "exception_message": null
  }
}

In [12]:
# 3 Harus dalam daftar tertentu
gdf.expect_column_values_to_be_in_set(
    "payment_method",
    ["Credit Card", "Debit Card", "PayPal", "Amazon Pay", "Gift Card"]
)


{
  "success": true,
  "result": {
    "element_count": 250,
    "missing_count": 0,
    "missing_percent": 0.0,
    "unexpected_count": 0,
    "unexpected_percent": 0.0,
    "unexpected_percent_total": 0.0,
    "unexpected_percent_nonmissing": 0.0,
    "partial_unexpected_list": []
  },
  "meta": {},
  "exception_info": {
    "raised_exception": false,
    "exception_traceback": null,
    "exception_message": null
  }
}

In [13]:

# 4 Tipe data harus sesuai
gdf.expect_column_values_to_be_in_type_list(
    "price", ["int64", "float64", "int32", "float32"]
)


{
  "success": true,
  "result": {
    "observed_value": "int64"
  },
  "meta": {},
  "exception_info": {
    "raised_exception": false,
    "exception_traceback": null,
    "exception_message": null
  }
}

In [14]:

# 5 Mean price antara rentang realistis
gdf.expect_column_mean_to_be_between("price", min_value=100, max_value=600)


{
  "success": true,
  "result": {
    "observed_value": 343.58,
    "element_count": 250,
    "missing_count": null,
    "missing_percent": null
  },
  "meta": {},
  "exception_info": {
    "raised_exception": false,
    "exception_traceback": null,
    "exception_message": null
  }
}

In [15]:

# 6 Maksimum total_sales tidak melebihi 6000
gdf.expect_column_max_to_be_between("total_sales", min_value=0, max_value=6000)


{
  "success": true,
  "result": {
    "observed_value": 6000,
    "element_count": 250,
    "missing_count": null,
    "missing_percent": null
  },
  "meta": {},
  "exception_info": {
    "raised_exception": false,
    "exception_traceback": null,
    "exception_message": null
  }
}

In [16]:

# 7 Status transaksi hanya boleh Pending, Completed, atau Cancelled
gdf.expect_column_values_to_be_in_set(
    "status", ["Pending", "Completed", "Cancelled"]
)


{
  "success": true,
  "result": {
    "element_count": 250,
    "missing_count": 0,
    "missing_percent": 0.0,
    "unexpected_count": 0,
    "unexpected_percent": 0.0,
    "unexpected_percent_total": 0.0,
    "unexpected_percent_nonmissing": 0.0,
    "partial_unexpected_list": []
  },
  "meta": {},
  "exception_info": {
    "raised_exception": false,
    "exception_traceback": null,
    "exception_message": null
  }
}

### Kesimpulan Hasil Great Expectations  
#### Dataset: `P2M3_haafizhhamda_data_clean.csv`

Dataset ini telah diuji menggunakan 7 expectation  untuk memastikan kualitas data sebelum proses transformasi atau analisis lanjutan.  
Hasil menunjukkan tingkat validitas data yang sangat baik (sekitar 98%).



---

- 1. `expect_column_values_to_be_unique("order_id")`
**Status:** Lulus sebagian  
**Deskripsi:** Semua baris memiliki nilai unik.

---

- 2. `expect_column_values_to_be_between("quantity", 1, 5)`
**Status:** Lulus  
**Deskripsi:** Semua nilai berada pada kisaran 1–5.  
**Kesimpulan:** Distribusi kuantitas pembelian realistis dan sesuai dengan perilaku transaksi normal.



---

- 3. `expect_column_values_to_be_in_set("payment_method", [...])`
**Status:** Lulus  
**Deskripsi:** Semua nilai metode pembayaran termasuk dalam daftar resmi:  
`["Credit Card", "Debit Card", "PayPal", "Amazon Pay", "Gift Card"]`  
**Kesimpulan:** Tidak ada metode pembayaran tidak dikenal, menandakan standarisasi data yang baik.

---

- 4. `expect_column_values_to_be_in_type_list("price", ["int64", "float64", ...])`
**Status:** Lulus  
**Deskripsi:** Semua nilai bertipe numerik (`float64`).  
**Kesimpulan:** Kolom `price` konsisten secara tipe data, tidak ada format string atau simbol mata uang.

---

- 5. `expect_column_mean_to_be_between("price", 100, 600)`
**Status:** Lulus  
**Deskripsi:** Nilai rata-rata harga sekitar 343.58 berada di rentang wajar.  
**Kesimpulan:** Harga produk stabil dan tidak menunjukkan anomali nilai ekstrem.

---

- 6. `expect_column_max_to_be_between("total_sales", 0, 6000)`
**Status:** Lulus  
**Deskripsi:** Nilai maksimum total_sales = 6000, sesuai batas atas.  
**Kesimpulan:** Tidak ditemukan nilai penjualan yang melebihi batas wajar.

---

- 7. `expect_column_values_to_be_in_set("status", ["Pending", "Completed", "Cancelled"])`
**Status:** Lulus  
**Deskripsi:** Semua nilai `status` sesuai dengan tiga kategori standar.  
**Kesimpulan:** Kolom `status` sudah terstandarisasi dengan baik dan konsisten untuk analisis transaksi.

---

### Ringkasan Umum
| Jenis Expectation | Jumlah | Status |
|--------------------|---------|---------|
| Wajib       | 4       | Lulus  |
| Tambahan           | 3       | Lulus semua |


---

### Kesimpulan Akhir
Secara keseluruhan, dataset **`P2M3_haafizhhamda_data_clean.csv`** memenuhi standar kualitas data yang baik:  
- Tidak terdapat nilai kosong, outlier ekstrem, atau kesalahan tipe data.  
- Struktur kolom lengkap dan konsisten.  
